cmd/xl-v1-multipart.go

```diff
@@ -375,8 +375,17 @@ func (xl xlObjects) PutObjectPart(ctx context.Context, bucket, object, uploadID
 	}
 
 	// Fetch buffer for I/O, returns from the pool if not allocates a new one and returns.
-	buffer := xl.bp.Get()
-	defer xl.bp.Put(buffer)
+	var buffer []byte
+	switch size := data.Size(); {
+	case size == 0:
+		buffer = make([]byte, 1) // Allocate atleast a byte to reach EOF
+	case size < blockSizeV1:
+		// No need to allocate fully blockSizeV1 buffer if the incoming data is smaller.
+		buffer = make([]byte, size)
```

Comment:
We should allocate 2*size here because the erasure coding will use [0:size] for data blocks and [size:2*size] for parity.
Otherwise we will allocate memory during erasure coding.

Reply:
This is what the bytepool does

```
        bp := bpool.NewBytePoolCap(setCount*drivesPerSet, blockSizeV1, blockSizeV1*2)
```

Reply:
Yes, so we should do the same for smaller buffers.